### PR TITLE
Support Configuration Cache

### PR DESCRIPTION
**Pull request by rpalcolea**
This should support config cache by following https://docs.gradle.org/8.4/userguide/configuration_cache.html#config_cache:requirements:external_processes when executing Git

All the integration tests apply the following now:

```
        new File(projectDir, 'gradle.properties') << '''org.gradle.configuration-cache=true'''.stripIndent()
```

which will enforce configuration cache during tests across all the specs

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -63,3 +63,21 @@ gradlePlugin {
         }
     }
 }
+
+/**
+ * This is temporary until config cache serialization is fixed when writing tests
+ * More in https://github.com/gradle/gradle/issues/25898
+ */
+tasks.withType(Test).configureEach {
```
Comment (rpalcolea) [inline build.gradle:71]: This is temporary until https://github.com/gradle/gradle/issues/25898 is addressed

File: src/integTest/groovy/nebula/plugin/release/GitVersioningIntegrationSpec.groovy [hunk 2]
```diff
@@ -48,6 +48,9 @@ abstract class GitVersioningIntegrationSpec extends IntegrationSpec {
 build/
 gradle.properties'''.stripIndent()
 
+        // Enable configuration cache :)
+        new File(projectDir, 'gradle.properties') << '''org.gradle.configuration-cache=true'''.stripIndent()
```
Comment (rpalcolea) [inline src/integTest/groovy/nebula/plugin/release/GitVersioningIntegrationSpec.groovy:52]: This enables config cache for all integration tests

File: src/main/groovy/nebula/plugin/release/git/base/ReleaseTask.groovy [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import nebula.plugin.release.git.command.GitWriteCommandsUtil
+import org.gradle.api.DefaultTask
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.TaskAction
+import org.gradle.work.DisableCachingByDefault
+
+@DisableCachingByDefault(because = "Publishing related tasks should not be cacheable")
```
Comment (OdysseusLives) [inline src/main/groovy/nebula/plugin/release/git/base/ReleaseTask.groovy:11]: 💯

File: src/main/groovy/nebula/plugin/release/git/base/ReleaseTask.groovy [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    @TaskAction
+    void release() {
+        GitWriteCommandsUtil gitCommands = gitWriteCommandsUtil.get()
+        String tagName = tagStrategy.get().maybeCreateTag(gitCommands, projectVersion.get())
+        if (tagName) {
+            logger.warn('Pushing changes in {} to {}', tagName, remote.get())
+            gitCommands.pushTag(remote.get(), tagName)
+        } else {
+            logger.warn('Nothing to push.')
```
Comment (OdysseusLives) [inline src/main/groovy/nebula/plugin/release/git/base/ReleaseTask.groovy:31]: Perhaps something like 
```suggestion
            logger.warn('No new tags to push for {}', remote.get())
```

File: src/main/groovy/nebula/plugin/release/git/command/GitReadCommand.groovy [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    abstract ExecOperations getExecOperations()
+
+    @CompileDynamic
+    String executeGitCommand(Object ... args) {
+        File rootDir = parameters.rootDir.get()
+        ByteArrayOutputStream output = new ByteArrayOutputStream()
+        ByteArrayOutputStream error = new ByteArrayOutputStream()
+        List<String> commandLineArgs = ["git", "--git-dir=${rootDir.absolutePath}/.git".toString(), "--work-tree=${rootDir.absolutePath}".toString()]
+        commandLineArgs.addAll(args)
+        execOperations.exec {
```
Comment (OdysseusLives) [inline src/main/groovy/nebula/plugin/release/git/command/GitReadCommand.groovy:27]: Nice! 🎉

File: src/main/groovy/nebula/plugin/release/git/command/GitReadCommand.groovy [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+ * These read only git commands use ValueSource approach for configuration cache
+ * @see {@link https://docs.gradle.org/8.4/userguide/configuration_cache.html#config_cache:requirements:external_processes}
+ */
+abstract class GitReadCommand implements ValueSource<String, GitCommandParameters> {
+
+    @Inject
+    abstract ExecOperations getExecOperations()
+
+    @CompileDynamic
+    String executeGitCommand(Object ... args) {
```
Comment (OdysseusLives) [inline src/main/groovy/nebula/plugin/release/git/command/GitReadCommand.groovy:21]: Love how this takes args

File: src/main/groovy/nebula/plugin/release/git/command/GitReadCommand.groovy [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+import org.gradle.process.ExecOperations
+
+import javax.inject.Inject
+import java.nio.charset.Charset
+
+/**
+ * These read only git commands use ValueSource approach for configuration cache
+ * @see {@link https://docs.gradle.org/8.4/userguide/configuration_cache.html#config_cache:requirements:external_processes}
+ */
+abstract class GitReadCommand implements ValueSource<String, GitCommandParameters> {
```
Comment (OdysseusLives) [inline src/main/groovy/nebula/plugin/release/git/command/GitReadCommand.groovy:15]: 💯 on documenting all these classes for usability and posterity

File: src/main/groovy/nebula/plugin/release/git/command/GitWriteCommandsUtil.groovy [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+import org.gradle.api.GradleException
+import org.gradle.process.ExecOperations
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import java.nio.charset.Charset
+
+/**
+ * These git commands involve creating or pushing tags
+ * Because these happen at task execution level, we are fine with not using ValueSource
```
Comment (OdysseusLives) [inline src/main/groovy/nebula/plugin/release/git/command/GitWriteCommandsUtil.groovy:13]: Great call-out! Thank you

**Review by OdysseusLives — APPROVED** [review]

**Pull request merged**